Make thread rng an optional feature

malramsay64 opened the pull request:
I'm not entirely sure this is useful to anyone else, however I thought it was and so thought I should offer the changes upstream. I have split all the changes into two pull requests with #26 being just the updates to rand and rand_core which is probably less objectionable, and this being the thread feature.

Running the tests with the thread feature requires the command
```sh
cargo test --features thread
```
which has been added to the CI

malramsay64 committed:
Update rand_core to 0.5 and rand to 0.7

This updates the versions of rand to the latest, making the required
updates to the codebase to support the newer features and style.

malramsay64 committed:
Make thread support optional

This is a refactoring such that the core sfmt random number generator
only depends on rand_core, with no dependency on rand. The rand crate is
required for the thread support though, so to ensure that this
functionality is still available I have included it as an optional
feature.

Removing the requirement of rand did additionally require that some of
the tests were modified to use the smaller range of functionality of the
rand_core crate.

malramsay64 committed:
Include thead feature in CI testing

This ensures that tests are run both with the thread feature enabled and
with the feature disabled.

termoshtt commented on Cargo.toml:
```diff
@@ -11,6 +11,9 @@ keywords = ["random", "rng"]
 license = "MIT"
 
+[features]
+thread = ["rand"]
```
I prefer `thread_rng` feature

termoshtt reviewed (changes requested):
Splitting the dependency to rand crate is a great idea, but I hope it is default feature.

```
default = ["thread_rng"]
thread_rng = ["rand"]
```

Could you rewrite tests with `--no-default-features` and `--features=thread_rng`?

malramsay64 committed:
Rename thread feature to thread_rng

The name thread_rng is a better descriptor of what the feature enables.

malramsay64 committed:
Make the thread_rng feature default

The thread_rng feature is now the default, which retains the support
from previous versions. Additionally the thread_rng is likely more
expected as a default feature.

malramsay64 commented:
Having thought about it, I agree that the thread_rng feature should be the default.

termoshtt commented:
Thanks!
